### PR TITLE
Pre-warmup RESTAPI endpoint

### PR DESCRIPTION
**Pull request by engahmeds3ed**
## Description

Add a new RESTAPI Endpoint to get the status of resource URLs from the Database.

## Response types

1. Fail when don't have the authority to do that action
```
{
    "code": "rest_forbidden",
    "message": "Sorry, you are not allowed to do that.",
    "data": {
        "status": 401
    }
}
```

2. Fail when RUCSS option is disabled.
```
{
    "success": false,
    "message": "Remove unused CSS option is disabled.",
    "data": []
}
```

3. Fail when pre-warmup process didn't start yet.
```
{
    "success": false,
    "message": "Pre-Warmup process didn't start yet.",
    "data": []
}
```

4. Success with the scan process & some wramed-up and some unwarmed resources
```
{
    "success": true,
    "data": {
        "scan_status": {
            "total_pages": 6,
            "scanned": 6,
            "fetched": 6
        },
        "warmup_status": {
            "total": 3,
            "warmed_count": 2,
            "notwarmed_resources": [
                "http://wordpress.localhost/wp-includes/js/dist/vendor/lodash.js?ver=4.17.19"
            ]
        }
    }
}
```

5. Success without any warmed-up resource yet.
```
{
    "success": true,
    "data": {
        "scan_status": {
            "total_pages": 6,
            "scanned": 6,
            "fetched": 6
        },
        "warmup_status": {
            "total": 3,
            "warmed_count": 0,
            "notwarmed_resources": [
                "http://wordpress.localhost/wp-includes/js/dist/vendor/wp-polyfill-fetch.js?ver=3.0.0",
                "http://wordpress.localhost/wp-includes/js/dist/vendor/wp-polyfill.js?ver=7.4.4",
                "http://wordpress.localhost/wp-content/themes/twentytwenty/assets/js/index.js?ver=1.2"
            ]
        }
    }
}
```

6. Success with all resources warmed-up
```
{
    "success": true,
    "data": {
        "scan_status": {
            "total_pages": 6,
            "scanned": 6,
            "fetched": 6
        },
        "warmup_status": {
            "total": 3,
            "warmed_count": 3,
            "notwarmed_resources": []
        }
    }
}
```

### CODE REVIEW

File: inc/Engine/Optimization/RUCSS/Warmup/Status/RESTWP.php [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+	 * @return void
+	 */
+	public function register_status_route() {
+		register_rest_route(
+			self::API_NAMESPACE,
+			self::API_ROUTE,
+			[
+				'methods'             => 'POST',
+				'callback'            => [ $this, 'respond_status' ],
+				'permission_callback' => [ $this, 'check_permissions' ],
```
Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Warmup/Status/RESTWP.php]: I would convert this to an anonymous method directly

File: inc/Engine/Optimization/RUCSS/Warmup/Status/RESTWP.php [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+	public function respond_status( WP_REST_Request $request ) : WP_REST_Response {
+		$urls = (array) $request->get_param( 'urls' );
+
+		if (
+			empty( $urls )
+			||
+			(bool) $this->options->get( 'remove_unused_css', 0 )
+		) {
+			return rest_ensure_response(
+				$this->return_error( __( 'Remove unused CSS option is disabled or Invalid request!', 'rocket' ) )
```
Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Warmup/Status/RESTWP.php]:
```suggestion
				$this->return_error( __( 'Remove unused CSS option is disabled or Invalid request.', 'rocket' ) )
```

File: inc/Engine/Optimization/RUCSS/Warmup/Status/RESTWP.php [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+			);
+		}
+
+		$warmup_total_resources_count = $this->resources_query->get_prewarmup_total_count( $urls );
+
+		if ( empty( $warmup_total_resources_count ) ) {
+			return rest_ensure_response(
+				$this->return_error( __( 'No resources into the DB!', 'rocket' ) )
```
Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Warmup/Status/RESTWP.php]:
```suggestion
				$this->return_error( __( 'No resources currently exists into the resources table.', 'rocket' ) )
```

File: inc/Engine/Optimization/RUCSS/Database/Queries/ResourcesQuery.php [hunk 4]
```diff
@@ -139,4 +139,57 @@ public function remove_by_url( $url ) {
...
 	}
+
+	/**
+	 * Get prewarmup total resources count.
+	 *
+	 * @param array $urls Urls to be checked.
+	 *
+	 * @return int
+	 */
+	public function get_prewarmup_total_count( array $urls ) : int {
```
Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Database/Queries/ResourcesQuery.php]:
```suggestion
	public function get_prewarmup_total_count( array $urls ): int {
```

File: inc/Engine/Optimization/RUCSS/Database/Queries/ResourcesQuery.php [hunk 5]
```diff
@@ -139,4 +139,57 @@ public function remove_by_url( $url ) {
...
+	}
+
+	/**
+	 * Get prewarmup warmed resources count.
+	 *
+	 * @param array $urls Urls to be checked.
+	 *
+	 * @return int
+	 */
+	public function get_prewarmup_warmed_count( array $urls ) : int {
```
Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Database/Queries/ResourcesQuery.php]:
```suggestion
	public function get_prewarmup_warmed_count( array $urls ): int {
```

File: inc/Engine/Optimization/RUCSS/Database/Queries/ResourcesQuery.php [hunk 6]
```diff
@@ -139,4 +139,57 @@ public function remove_by_url( $url ) {
...
+	}
+
+	/**
+	 * Get prewarmup NOT warmed resources' urls.
+	 *
+	 * @param array $urls Urls to be checked.
+	 *
+	 * @return array|string[]
+	 */
+	public function get_prewarmup_notwarmed_urls( array $urls ) : array {
```
Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Database/Queries/ResourcesQuery.php]:
```suggestion
	public function get_prewarmup_notwarmed_urls( array $urls ): array {
```

File: inc/Engine/Optimization/RUCSS/Warmup/Status/RESTWP.php [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+	/**
+	 * Generates the CPCSS for the requested post ID.
+	 *
+	 * @since 3.6
+	 *
+	 * @param WP_REST_Request $request WP REST request response.
+	 *
+	 * @return WP_REST_Response
+	 */
+	public function respond_status( WP_REST_Request $request ) : WP_REST_Response {
```
Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Warmup/Status/RESTWP.php]:
```suggestion
	public function respond_status( WP_REST_Request $request ): WP_REST_Response {
```

File: inc/Engine/Optimization/RUCSS/Warmup/Status/RESTWP.php [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+	 * Get array of error response.
+	 *
+	 * @since 3.9
+	 *
+	 * @param string $message Error message.
+	 * @param array  $data Data to be passed.
+	 *
+	 * @return array
+	 */
+	private function return_error( string $message, array $data = [] ) : array {
```
Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Warmup/Status/RESTWP.php]:
```suggestion
	private function return_error( string $message, array $data = [] ): array {
```

File: inc/Engine/Optimization/RUCSS/Warmup/Status/RESTWP.php [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+	/**
+	 * Get array of success response.
+	 *
+	 * @since 3.9
+	 *
+	 * @param array $data Data to be passed.
+	 *
+	 * @return array
+	 */
+	private function return_success( array $data = [] ) : array {
```
Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Warmup/Status/RESTWP.php]:
```suggestion
	private function return_success( array $data = [] ): array {
```

**Review by remyperona — CHANGES_REQUESTED** [review]

File: inc/Engine/Optimization/RUCSS/Warmup/Status/RESTWP.php (hunk 1 above)
Comment (engahmeds3ed) [reply inc/Engine/Optimization/RUCSS/Warmup/Status/RESTWP.php]: u mean check_permissions one?

Comment (remyperona) [reply inc/Engine/Optimization/RUCSS/Warmup/Status/RESTWP.php]: yes, like the following:
 ```
'permission_callback' => function() {
    return current_user_can( 'rocket_remove_unused_css' );
},
```

Comment (engahmeds3ed) [reply inc/Engine/Optimization/RUCSS/Warmup/Status/RESTWP.php]: yes, I know the anonymous function for sure but really I hate them :D for many reasons and one of them is to add docblock but it's ok with this one-line method.

**Comment by remyperona** [comment]
The endpoint also needs to give data about the first scanner status, taking information from the `wp_rocket_resources_scanner` option, which contains an array with items like the following:
```
[
'url' => 'url',
'type' => 'post_type',
'is_scanned' => 'true',
]
```

**Review by remyperona — APPROVED** [review]

**Pull request merged**